### ISSUE TITLE
Segmentation Fault when running grad() -- invalid address reference

### ISSUE THREAD

**Issue by jinze1994** [stan-dev/math#868]
#### Summary:
Segmentation Fault when running grad(): invalid address reference

#### Description:
I am trying to implement a two-layer neural network by stan-math (All the test codes is attached in Additional information.) After compiling successfully, I encounter the segmentation fault. Then I use gdb and find out the invalid address reference, and I guess it's because of the wrong C++ pointer modification. Please fix this bug asap.

#### Reproducible Steps:
Compile and run the code in Additional information.
```
$ g++ test.cpp -g
$ gdb ./a.out
(gdb) r
Starting program: /home/jinze/Code/stan/a.out 
0 0.918244
Program received signal SIGSEGV, Segmentation fault.
0x00005555555a4546 in stan::math::multiply_mat_vari<double, -1, -1, stan::math::var, -1>::chain (
    this=0x5555557f4e70) at /usr/local/include/stan/math/rev/mat/fun/multiply.hpp:282
282	      variRefB_[i]->adj_ += adjB(i);
(gdb) bt
#0  0x00005555555a4546 in stan::math::multiply_mat_vari<double, -1, -1, stan::math::var, -1>::chain
    (this=0x5555557f4e70) at /usr/local/include/stan/math/rev/mat/fun/multiply.hpp:282
#1  0x0000555555558851 in stan::math::grad (vi=0x7ffff1f6f170)
    at /usr/local/include/stan/math/rev/core/grad.hpp:44
#2  0x000055555555c141 in stan::math::var::grad (this=0x7fffffffdeb0)
    at /usr/local/include/stan/math/rev/core/var.hpp:331
#3  0x000055555555a116 in main () at test.cpp:79
(gdb) p i
$1 = 2
(gdb) p variRefB_[i]
$2 = (stan::math::vari *) 0xfff6a54c7800007f
(gdb) p variRefB_[i]->adj_
Cannot access memory at address 0xfff6a54c7800008f
```
We can see that the Segentation Fault is owing to the wrong address wich `variRefB_[i]` contains. Furthermore, we can track this variable from the constructor function:
```
(gdb) l /usr/local/include/stan/math/rev/mat/fun/multiply.hpp:266
261	      Ad_[i] = A.coeffRef(i);
262	    for (size_type i = 0; i < B.size(); ++i) {
263	      variRefB_[i] = B.coeffRef(i).vi_;
264	      Bd_[i] = B.coeffRef(i).val();
265	    }
266	    MatrixXd AB = Map<MatrixXd>(Ad_, A_rows_, A_cols_)
267	                  * Map<MatrixXd>(Bd_, A_cols_, B_cols_);
268	    for (size_type i = 0; i < AB.size(); ++i)
269	      variRefAB_[i] = new vari(AB.coeffRef(i), false);
270	  }
(gdb) b /usr/local/include/stan/math/rev/mat/fun/multiply.hpp:282
Breakpoint 1 at 0x5555555a450d: file /usr/local/include/stan/math/rev/mat/fun/multiply.hpp, line 282.
(gdb) p variRefB_[2]
$3 = (stan::math::vari *) 0x5555557f4eb8
(gdb) p *(variRefB_[2])
$4 = {_vptr.vari = 0x7ffff4296010, val_ = -0.1678120248923195, adj_ = 0}
(gdb) delete 1
(gdb) c
Continuing.

Program received signal SIGSEGV, Segmentation fault.
0x00005555555a4546 in stan::math::multiply_mat_vari<double, -1, -1, stan::math::var, -1>::chain (
    this=0x5555557f4e70) at /usr/local/include/stan/math/rev/mat/fun/multiply.hpp:282
282	      variRefB_[i]->adj_ += adjB(i);
(gdb) p variRefB_[2]
$5 = (stan::math::vari *) 0xfff6a54c7800007f
```
We can see that the `variRefB_[2]` has been corrupted by something else. `watch` might be helpfu.

Thank you for your reading, please fix this bug asap.

#### Current Output:
Segmentation Fault

#### Expected Output:
No Segmentation Fault and corrupt pointer.

#### Additional Information:
Provide any additional information here.
```C++
#include <cmath>                                                                                     
#include <boost/random.hpp>
#include <boost/random/normal_distribution.hpp>
#include <iostream>
#include <stan/math.hpp>

#define TRAIN_N 55000
#define VALID_N 5000
#define TEST_N 10000

#define IMAGE_ROW 28
#define IMAGE_COL 28
#define IMAGE_LEN (IMAGE_ROW * IMAGE_COL)
#define LABEL_LEN 10

typedef Eigen::Matrix<stan::math::var, Eigen::Dynamic, Eigen::Dynamic> Matrix;
typedef Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1> Vector;
typedef Eigen::Matrix<stan::math::var, 1, Eigen::Dynamic> RowVector;

stan::math::var sample(stan::math::var dummy) {
  static boost::mt19937 rng;
  static boost::normal_distribution<> nd(0.0, 0.1);
  return nd(rng);
}

stan::math::var update(stan::math::var old) {
  double grad = old.adj();
  grad = std::max(std::min(grad, 5.0), -5.0);
  return stan::math::var(old.val() - 0.01 * grad);
}

stan::math::var relu(stan::math::var old) {
  return stan::math::fmax(old, 0.0);
}

Vector softmax_cross_entropy(
    Matrix& logits,
    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> labels) {
  int batch_size = logits.rows();
  Vector v(batch_size);
  for (int i=0; i<batch_size; i++)
    v(i) = -stan::math::dot_product(stan::math::log_softmax(stan::math::row(logits, i+1)),
                                    stan::math::row(labels, i+1));
  return v;
}

struct MnistBatch {
  Eigen::MatrixXd images;
  Eigen::MatrixXd labels;
};

int main() {
  int batch_size = 128;
  int hidden_size = 1024;
  clock_t begin = clock();

  Matrix W1 = Matrix::Zero(IMAGE_LEN, hidden_size).unaryExpr(std::ptr_fun(sample));
  Vector b1 = Matrix::Zero(hidden_size, 1);
  Matrix W2 = Matrix::Zero(hidden_size, 10).unaryExpr(std::ptr_fun(sample));
  Vector b2 = Matrix::Zero(10, 1);
  MnistBatch batch;

  Matrix h1;
  Matrix h2;

  for (int i=0; i<=2000; i++) {
    batch.images = Eigen::MatrixXd::Random(batch_size, IMAGE_LEN);
    batch.labels = Eigen::MatrixXd::Random(batch_size, LABEL_LEN);

    stan::math::recover_memory();

    h1 = (stan::math::multiply(batch.images, W1) + stan::math::rep_matrix(stan::math::transpose(b1), batch.images.rows())).unaryExpr(std::ptr_fun(relu));
    h2 = stan::math::multiply(h1, W2) + stan::math::rep_matrix(stan::math::transpose(b2), h1.rows());
    stan::math::var loss = stan::math::mean(softmax_cross_entropy(h2, batch.labels));

    double cur_loss = loss.val();
    std::cout << i << " " << cur_loss << std::endl;
    stan::math::set_zero_all_adjoints();
    loss.grad();
    W1 = W1.unaryExpr(std::ptr_fun(update));
    b1 = b1.unaryExpr(std::ptr_fun(update));
    W2 = W2.unaryExpr(std::ptr_fun(update));
    b2 = b2.unaryExpr(std::ptr_fun(update));
  }
}
```

#### Current Version:
v2.17.1

**Comment by syclik** [comment]
Thanks for submitting an issue.

I'm responding on my phone, so this may not be completely correct. I believe this isn't a Math issue, but a misuse of the library. 

The construction of W1, b1, W2, and b2 should happen *after* the call to recover_memory(). That call resets the autodiff stack and these variables are no longer valid.

**Comment by jinze1994** [comment]
Is there any light-weighted recover function to rebuild the stack or to re-evaluate the loss after changing the weight, without recover memory and copy the weight from double array to stan::var array?

**Comment by syclik** [comment]
No, there isn't. We can't reuse the autodiff stack (also called reusing the tape in the literature) because we're eagerly building parts of the chain rule as we're doing the forward sweep. More details are in the arXiv paper: https://arxiv.org/abs/1509.07164
